x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/v2/TransportRollupV2Action.java

```diff
@@ -0,0 +1,140 @@
+            // update Rollup metadata to include this index
+            clusterService.submitStateUpdateTask("update-rollup-metadata", new ClusterStateUpdateTask() {
+
+                @Override
+                public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                    listener.onResponse(new RollupV2Action.Response(true));
+                }
+
+                @Override
+                public ClusterState execute(ClusterState currentState) throws Exception {
+                    String rollupIndexName = rollupV2Task.config().getRollupIndex();
+                    IndexMetadata rollupIndexMetadata = currentState.getMetadata().index(rollupIndexName);
+                    Index rollupIndex = rollupIndexMetadata.getIndex();
+                    // TODO(talevy): find better spot to get the original index name
+                    // extract created rollup index original index name to be used as metadata key
+                    String originalIndexName = rollupV2Task.config().getSourceIndex();
+                    Map<String, String> idxMetadata = currentState.getMetadata().index(originalIndexName)
+                        .getCustomData(RollupMetadata.TYPE);
+                    String rollupGroupKeyName = (idxMetadata == null) ?
+                        originalIndexName : idxMetadata.get(RollupMetadata.SOURCE_INDEX_NAME_META_FIELD);
+                    Map<String, String> rollupIndexRollupMetadata = new HashMap<>();
+                    rollupIndexRollupMetadata.put(RollupMetadata.SOURCE_INDEX_NAME_META_FIELD, rollupGroupKeyName);
+                    final RollupMetadata rollupMetadata = currentState.metadata().custom(RollupMetadata.TYPE);
+                    final Map<String, RollupGroup> rollupGroups;
+                    if (rollupMetadata == null) {
+                        rollupGroups = new HashMap<>();
+                    } else {
+                        rollupGroups = new HashMap<>(rollupMetadata.rollupGroups());
+                    }
+                    DateHistogramGroupConfig dateConfig = rollupV2Task.config().getGroupConfig().getDateHistogram();
+                    WriteableZoneId rollupDateZoneId = WriteableZoneId.of(dateConfig.getTimeZone());
+                    if (rollupGroups.containsKey(rollupGroupKeyName)) {
+                        RollupGroup group = rollupGroups.get(rollupGroupKeyName);
+                        group.add(rollupIndexName, dateConfig.getInterval(), rollupDateZoneId);
+                    } else {
+                        List<String> indices = new ArrayList<>();
+                        Map<String, DateHistogramInterval> intervals = new HashMap<>();
+                        Map<String, WriteableZoneId> timezones = new HashMap<>();
+                        indices.add(rollupIndexName);
+                        intervals.put(rollupIndexName, dateConfig.getInterval());
+                        timezones.put(rollupIndexName, rollupDateZoneId);
```

Comment:
We're bleeding implementation details here a bit.  I feel like we should just create an empty `RollupGroup` and then use the `RollupGroup#add(indexName, dateInterval, ZoneId)` method you used in the other branch of this conditional.  Seems weird to me that this method decides the concrete map implementation the `RollupGroup` is using, for example.

Reply:
good point. I'll work to clean this up